Text2SQL using Hugging Face Datasets Server API and Motherduck DuckDB-NSQL-7B (#1945)
huggingface/blog: duckdb-nsql-7b.md

In recent months, significant strides have been made in this arena. [MotherDuck](https://motherduck.com/) and [Numbers Station](https://numbersstation.ai/) unveiled their latest innovation: [DuckDB-NSQL-7B](https://huggingface.co/motherduckdb/DuckDB-NSQL-7B-v0.1), a state-of-the-art LLM designed specifically for [DuckDB SQL](https://duckdb.org/). What is this model’s mission? To empower users with the ability to unlock insights from their data effortlessly.

Initially fine-tuned from Meta’s original [Llama-2–7b](https://huggingface.co/meta-llama/Llama-2-7b) model using a broad dataset covering general SQL queries, DuckDB-NSQL-7B underwent further refinement with DuckDB text-to-SQL pairs. Notably, its capabilities extend beyond crafting `SELECT` statements; it can generate a wide range of valid DuckDB SQL statements, including official documentation and extensions, making it a versatile tool for data exploration and analysis.

In this article, we will learn how to deal with text2sql tasks using the DuckDB-NSQL-7B model, Hugging Face datasets server API for parquet files and duckdb for data retrieval.

<p align="center">

>  <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/duckdb-nsql-7b/text2sql-flow.png" alt="text2sql flow" style="width: 90%; height: auto;"><br>

Reviewer: Maybe edit the image to not have the components touch the image borders and leave some space ?

Because for users in dark mode they'll see something like this, with the dark borders touching the top rectangle for example.

<img width="866" alt="image" src="https://github.com/huggingface/blog/assets/42851186/8a3e1978-e9c9-433e-a7cb-afdbcdf2a981">

Author: Sure! I will modify the image before merging

Author: Done, https://huggingface.co/datasets/huggingface/documentation-images/discussions/313 once merged, I will merge this PR as well :)

Reviewer: Nice thanks :)

Reviewer 2: Great eye there @lhoestq 🙌 Just fyi, I think the blog does not support dark mode yet IIRC

Reviewer: (I just checked it does support dark mode ;) )

Reviewer 2: lol, my test was wrong then. Sorry!

Author: btw @pcuenca  can you help me merge https://huggingface.co/datasets/huggingface/documentation-images/discussions/313? So that I can merge this PR as well (I don't have permission to do it in huggingface/documentation-images) thanks in advance!

Reviewer 2: Done, sorry

---
title: "Text2SQL using Hugging Face Datasets Server API and Motherduck DuckDB-NSQL-7B" 
thumbnail: /blog/assets/duckdb-nsql-7b/thumbnail.png
authors:

> - user: asoria

Reviewer 3: ```suggestion
- user: asoria, tdoehmen, senwu, lorr
```

> - user: asoria, tdoehmen, senwu, lorr

Reviewer 2: ```suggestion
- user: asoria
- user: tdoehmen
  guest: true
- user: senwu
  guest: true
- user: lorr
  guest: true
```

(Sorry, your commit got in while I was reviewing)

This is where groundbreaking advancements in AI, LLM technology, step in to bridge the divide. Imagine conversing with your data effortlessly, simply stating your information needs in plain language and having the model translate your request into a query. 

In recent months, significant strides have been made in this arena. [MotherDuck](https://motherduck.com/) and [Numbers Station](https://numbersstation.ai/) unveiled their latest innovation: [DuckDB-NSQL-7B](https://huggingface.co/motherduckdb/DuckDB-NSQL-7B-v0.1), a state-of-the-art LLM designed specifically for [DuckDB SQL](https://duckdb.org/). What is this model’s mission? To empower users with the ability to unlock insights from their data effortlessly.

Initially fine-tuned from Meta’s original [Llama-2–7b](https://huggingface.co/meta-llama/Llama-2-7b) model using a broad dataset covering general SQL queries, DuckDB-NSQL-7B underwent further refinement with DuckDB text-to-SQL pairs. Notably, its capabilities extend beyond crafting `SELECT` statements; it can generate a wide range of valid DuckDB SQL statements, including official documentation and extensions, making it a versatile tool for data exploration and analysis.

> In this article, we will learn how to create deal with text2sql task using DuckDB-NSQL-7B model, Hugging Face datasets server API for parquet files and duckdb for data retrieval.

Reviewer 3: ```suggestion
In this article, we will learn how to deal with text2sql tasks using the DuckDB-NSQL-7B model, Hugging Face datasets server API for parquet files and duckdb for data retrieval.
```

llama = Llama(
       model_path="DuckDB-NSQL-7B-v0.1-q8_0.gguf", # Path to local model
       n_gpu_layers=-1,
)
```

The main goal of `llama.cpp` is to enable LLM inference with minimal setup and state-of-the-art performance on a wide variety of hardware - locally and in the cloud, we will use this approach.

#### Hugging Face Datasets Server API for more than 120K datasets

Data is a crucial component in any Machine Learning endeavor. Hugging Face is a valuable resource, offering access to over 120,000 free and open datasets spanning various formats, including CSV, Parquet, JSON, audio, and image files.

Each dataset hosted by Hugging Face comes equipped with a comprehensive dataset viewer. This viewer provides users essential functionalities such as statistical insights, data size assessment, full-text search capabilities, and efficient filtering options. This feature-rich interface empowers users to easily explore and evaluate datasets, facilitating informed decision-making throughout the machine learning workflow.

For this demo, we will be using the [world-cities-geo](https://huggingface.co/datasets/jamescalam/world-cities-geo) dataset.

<p align="center">

>  <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/duckdb-nsql-7b/dataset-viewer.png" alt="dataset viewer" style="width: 90%; height: auto;"><br>

Author: Sure, I will do it

>  <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/duckdb-nsql-7b/dataset-viewer.png" alt="dataset viewer" style="width: 90%; height: auto;"><br>

Author: Done. I changed the images for `download.png` (downloading the model) and `dataset-viewer.png`, which were the two ones with a low-quality https://huggingface.co/datasets/huggingface/documentation-images/discussions/311.

>  <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/duckdb-nsql-7b/dataset-viewer.png" alt="dataset viewer" style="width: 90%; height: auto;"><br>

Reviewer 2: Thank you!

- Get the dataset **size** (in number of rows or bytes)
- Download and view **rows at any index** in the dataset
- **Search** a word in the dataset
- **Filter** rows based on a query string
- Get insightful **statistics** about the data
- Access the dataset as **parquet files** to use in your favorite processing or analytics framework

In this demo, we will use the last functionality, auto-converted parquet files.

#### Generate SQL queries based on a given text for your Hugging Face Dataset

First, [download](https://huggingface.co/motherduckdb/DuckDB-NSQL-7B-v0.1-GGUF/blob/main/DuckDB-NSQL-7B-v0.1-q8_0.gguf) the quantized models version of DuckDB-NSQL-7B-v0.1

<p align="center">
 <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/duckdb-nsql-7b/download.png" alt="download model" style="width: 90%; height: auto;"><br>
<em>Downloading the model</em>
</p>

Alternatively, you can execute the following code:

```
huggingface-cli download motherduckdb/DuckDB-NSQL-7B-v0.1-GGUF DuckDB-NSQL-7B-v0.1-q8_0.gguf --local-dir . --local-dir-use-symlinks False
```

Now, lets install the needed dependencies:

```

> CMAKE_ARGS="-DLLAMA_METAL=on" pip install llama-cpp-python

Author: My mistake, actually all the code is running on an Ubuntu machine so no needed, I just copied/pasted it from the doc. I will remove it since it is not needed (neither in the colab notebook).

pip install duckdb
```

> For the text2SQL model, we will use a prompt with the following structure:

Reviewer 3: ```suggestion
For the text-to-SQL model, we will use a prompt with the following structure:
```

```
   ### Instruction:
   Your task is to generate valid duckdb SQL to answer the following question.
   ### Input:
   Here is the database schema that the SQL query will run on:
   {ddl_create}
  
   ### Question:
   {query_input}
   ### Response (use duckdb shorthand if possible):
```

- **ddl_create** will be the dataset schema as a SQL `CREATE` command
- **query_input** will be the user in a natural language way

So, we need to tell to the model about the schema of the Hugging Face dataset, for that, we are going to do a trick:

> - Get the URL of the first parquet file of the dataset hosted in `refs/convert/parquet` in Hugging Face. See https://huggingface.co/docs/datasets-server/parquet

Reviewer 4: instead of these three points, we could "just" get https://huggingface.co/api/datasets/jamescalam/world-cities-geo/parquet

```json
{"default":{"train":["https://huggingface.co/api/datasets/jamescalam/world-cities-geo/parquet/default/train/0.parquet"]}}
```

or even quicker, directly use https://huggingface.co/api/datasets/jamescalam/world-cities-geo/parquet/default/train/0.parquet

Author: Done, but I kept the final part where it mentioned that the parquet files are hosted under refs/convert/parquet branch in the hub to have the auto convert feature visibility (I can remove it if needed).

```
GET https://datasets-server.huggingface.co/parquet?dataset=jamescalam/world-cities-geo
```

```
{
   "parquet_files":[
      {
         "dataset":"jamescalam/world-cities-geo",
         "config":"default",
         "split":"train",
         "url":"https://huggingface.co/datasets/jamescalam/world-cities-geo/resolve/refs%2Fconvert%2Fparquet/default/train/0000.parquet",
         "filename":"0000.parquet",
         "size":499223
      }
   ],
   "pending":[
      
   ],
   "failed":[
      
   ],
   "partial":false
}
```

The [parquet file](https://huggingface.co/datasets/jamescalam/world-cities-geo/resolve/refs%2Fconvert%2Fparquet/default/train/0000.parquet) is hosted in Hugging Face viewer under `refs/convert/parquet` revision:

<p align="center">
 <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/duckdb-nsql-7b/parquet.png" alt="parquet file" style="width: 90%; height: auto;"><br>
<em>Parquet file</em>
</p>

- Simulate a [DuckDB](https://duckdb.org/) table creation from the first row of the parquet file

```
import duckdb
con = duckdb.connect()

> con.execute("INSTALL 'httpfs'; LOAD httpfs;")

Author: Thanks! I removed it

con.execute(f"CREATE TABLE data as SELECT * FROM '{first_parquet_url}' LIMIT 1;")

result = con.sql("SELECT sql FROM duckdb_tables() where table_name ='data';").df()
ddl_create = result.iloc[0,0]
con.close()
```

The `CREATE` schema DDL is:

```
CREATE TABLE "data"(
    city VARCHAR, 
    country VARCHAR, 
    region VARCHAR,
    continent VARCHAR, 
    latitude DOUBLE, 
    longitude DOUBLE, 
    x DOUBLE, 
    y DOUBLE, 
    z DOUBLE
);
```

> And, as you can see, it matches the columns as in the dataset viewer:

Reviewer 2: ```suggestion

And, as you can see, it matches the columns in the dataset viewer:
```

<p align="center">
 <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/duckdb-nsql-7b/columns.png" alt="dataset columns" style="width: 90%; height: auto;"><br>
<em>Dataset columns</em>
</p>

- Now, we can construct the prompt with the **ddl_create** and the **query** input

```
prompt = """### Instruction:
   Your task is to generate valid duckdb SQL to answer the following question.
   ### Input:
   Here is the database schema that the SQL query will run on:
   {ddl_create}
  
   ### Question:
   {query_input}
   ### Response (use duckdb shorthand if possible):
   """
```
If the user wants to know the **Cities from Albania country**, the prompt will look like this:

```
query = "Cities from Albania country"
prompt = prompt.format(ddl_create=ddl_create, query_input=query)
```

> 

Reviewer 2: ```suggestion

So the expanded prompt that will be sent to the LLM looks like this:
```

```
### Instruction:
Your task is to generate valid duckdb SQL to answer the following question.

### Input:
Here is the database schema that the SQL query will run on:
CREATE TABLE "data"(city VARCHAR, country VARCHAR, region VARCHAR, continent VARCHAR, latitude DOUBLE, longitude DOUBLE, x DOUBLE, y DOUBLE, z DOUBLE);
  
### Question:
Cities from Albania country

### Response (use duckdb shorthand if possible):
```

- It is time to send the prompt to the model

```
from llama_cpp import Llama

llm = Llama(
       model_path="DuckDB-NSQL-7B-v0.1-q8_0.gguf",
       n_ctx=2048,
       n_gpu_layers=50
   )
pred = llm(prompt, temperature=0.1, max_tokens=1000)
sql_output = pred["choices"][0]["text"]

```

The output SQL command will point to a `data` table, but since we don't have a real table but just a reference to the parquet file, we will replace all `data` occurrences by the `first_parquet_url`:

```
sql_output = sql_output.replace("FROM data", f"FROM '{first_parquet_url}'")
```

And the final output will be:

```

> SELECT city FROM 'https://huggingface.co/datasets/jamescalam/world-cities-geo/resolve/refs%2Fconvert%2Fparquet/default/train/0000.parquet' WHERE country = 'Albania'

Reviewer 4: same here:

```
SELECT city FROM 'https://huggingface.co/api/datasets/jamescalam/world-cities-geo/parquet/default/train/0.parquet' WHERE country = 'Albania'
```

<p align="center">
 <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/duckdb-nsql-7b/search.png" alt="search result" style="width: 90%; height: auto;"><br>
<em>Search result for Albania country</em>
</p>

You can also get the same result calling directly to the search or filter API:

- Using [/search](https://huggingface.co/docs/datasets-server/search?code=python#search-text-in-a-dataset) API

```
import requests
API_URL = "https://datasets-server.huggingface.co/search?dataset=jamescalam/world-cities-geo&config=default&split=train&query=Albania"
def query():
    response = requests.get(API_URL)
    return response.json()
data = query()
```

- Using [filter](https://huggingface.co/docs/datasets-server/filter) API
```
import requests
API_URL = "https://datasets-server.huggingface.co/filter?dataset=jamescalam/world-cities-geo&config=default&split=train&where=country='Albania'"
def query():
    response = requests.get(API_URL)
    return response.json()
data = query()
```

Our final demo will be a Hugging Face space that looks like this:

> <figure class="image table text-center m-0 w-full">

Reviewer 2: Just fyi, you can also embed the demo in the blog post if you want, see for example https://github.com/huggingface/blog/blob/0d6df18b5a648db9492dea3fe98ee22061d216f2/leaderboard-upstage.md?plain=1#L21

Author: Thanks! I didn't know it was possible to embed a space.
I am not sure if it is only me, but when trying to add something like that in my article, it gives `Connection errorred out`. Also, I see that the article gives that error; maybe an error in the gradio library?
![image](https://github.com/huggingface/blog/assets/5564745/9be97376-c626-4273-9054-82aa1797f24a)

Anyway, I would like to keep the mp4 for this time to specifically showcase how the demo runs while generating the SQL text and showing the result data frame.

Reviewer 2: Interesting, that was not happening before! Yes, may be related with the library version.
